hrqb/base/task.py

```diff
@@ -135,9 +135,20 @@ def dwclient(self) -> DWClient:
         return DWClient()  # pragma: nocover
 
     @property
-    def sql_query(self) -> str | None:
-        """SQL query from string to execute."""
-        return None
+    def sql_query(self) -> str:
+        """SQL query from string to execute.
+
+        Default behavior is to read a SQL file defined by self.sql_file.  Or, this
+        property can be overridden to provide a SQL query explicitly.
+        """
+        if not self.sql_file:
+            message = (
+                "Property 'sql_file' must be set or property 'sql_query' overridden to "
```

Comment:
Thanks for digging into this with me.

Counter-point: as a base Luigi task, it's designed to be extended when we define new task classes that extend this base class, and not set directly.

Here is an example coming from a real extract task coming in future PRs:

```python
class ExtractDWEmployeeAppointments(SQLQueryExtractTask):
    """Query Data Warehouse for employee appointment data."""

    pipeline = luigi.Parameter()
    stage = luigi.Parameter("Extract")

    @property
    def sql_file(self) -> str:
        return "hrqb/tasks/sql/employee_appointments.sql"
```

Hoping this demonstrates that it's quite normal to override a property in this way, and not set it directly.

And [here is an example from the testing fixtures](https://github.com/MITLibraries/hrqb-client/pull/16/files#diff-52a8ed9fa9fb3bfc1b3996f5eb619d05ffd78905a66250ae543c4054865eaefaR49-R51).

I take your point that manually setting a property to override it is somewhat inconvenient in testing, but I think it's more analogous to overrding a method with default behavior via class inheritance.

And while it's _kind_ of like an abstract method, it technically does not need to be defined.  It would also be acceptable to just define `sql_query` and provide the SQL query explicitly, like [this test fixture demonstrates](https://github.com/MITLibraries/hrqb-client/pull/16/files#diff-52a8ed9fa9fb3bfc1b3996f5eb619d05ffd78905a66250ae543c4054865eaefaR65-R69).

Thoughts?  You okay with proceeding with this?

Reply:
FWIW, I had considered making it a `luigi.Parameter()`, and I suppose it still could be, but I don't think that made sense for the `sql_query` parameter, and thought the symmetry was ultimately a better fit.

Ultimately, these base classes are for convenience when defining new classes, but I think it's implied (or it should be) that when defining classes, whatever works and makes sense in that instance is okay.

Reply:
@ghukill Ah, thanks for pointing out the test fixture! Yep, sounds good to me.